File: Python/intersection_point_of_two_Linked_Lists.py
# Python program to get intersection
# point of two linked list
class Node :
	def __init__(self, d):
		self.data = d;
		self.next = None;

# Function to print the list
def Print(n):
	cur = n;
	while (cur != None) :
		print(cur.data, end=" ");
		cur = cur.next;
	print("");

# Function to find the intersection of two node
def MegeNode(n1, n2):
	
	# Define hashset
	hs = set();

	while (n1 != None):
		hs.add(n1);
		n1 = n1.next;
	while (n2 != None):
		if (n2 in hs):
			return n2;
		n2 = n2.next;
	
	return None;


# Driver code

# list 1
n1 = Node(1);
n1.next = Node(2);
n1.next.next = Node(3);
n1.next.next.next = Node(4);
n1.next.next.next.next = Node(5);
n1.next.next.next.next.next = Node(6);
n1.next.next.next.next.next.next = Node(7);

# list 2
n2 = Node(10);
n2.next = Node(9);
n2.next.next = Node(8);
n2.next.next.next = n1.next.next.next;

Print(n1);
Print(n2);

print(MegeNode(n1, n2).data);

File: Java/Data Structures/intersection_point_of_two_Linked_Lists.java
// Java Program to get intersection point of two linked
// lists.

class GFG {

	static class Node {
		int data;
		Node next;
		Node(int d)
		{
			data = d;
			next = null;
		}
	}

	/* function to get the intersection point of two linked
	lists head1 and head2 */
	public Node getIntersectionNode(Node head1, Node head2)
	{
		while (head2 != null) {
			Node temp = head1;
			while (temp != null) {
				// if both Nodes are same
				if (temp == head2) {
					return head2;
				}
				temp = temp.next;
			}
			head2 = head2.next;
		}
		// If intersection is not present between the lists,
		// return NULL.
		return null;
	}

	public static void main(String[] args)
	{
		GFG list = new GFG();

		Node head1, head2;

		/*
				Create two linked lists

				1st 3->6->9->15->30
				2nd 10->15->30

				15 is the intersection point
		*/

		head1 = new Node(10);
		head2 = new Node(3);

		Node newNode = new Node(6);
		head2.next = newNode;

		newNode = new Node(9);
		head2.next.next = newNode;

		newNode = new Node(15);
		head1.next = newNode;
		head2.next.next.next = newNode;

		newNode = new Node(30);
		head1.next.next = newNode;

		head1.next.next.next = null;

		Node intersectionPoint
			= list.getIntersectionNode(head1, head2);

		if (intersectionPoint == null) {
			System.out.print(" No Intersection Point \n");
		}
		else {
			System.out.print("Intersection Point: "
							+ intersectionPoint.data);
		}
	}
}
